aptos-move/framework/src/natives/transaction_context.rs

```diff
@@ -34,6 +42,68 @@ impl NativeTransactionContext {
     }
 }
 
+/***************************************************************************************************
+ * native fun get_txn_hash
+ *
+ *   gas cost: base_cost
+ *
+ **************************************************************************************************/
+#[derive(Clone, Debug)]
+pub struct GetTxnHashGasParameters {
+    pub base: InternalGas,
+}
+
+fn native_get_txn_hash(
+    gas_params: &GetTxnHashGasParameters,
+    context: &mut SafeNativeContext,
+    mut _ty_args: Vec<Type>,
+    _args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    context.charge(gas_params.base)?;
+
+    let transaction_context = context.extensions().get::<NativeTransactionContext>();
+
+    Ok(smallvec![Value::vector_u8(
+        transaction_context.txn_hash.clone()
+    )])
+}
+
+/***************************************************************************************************
+ * native fun create_uuid
+ *
+ *   gas cost: base_cost
+ *
+ **************************************************************************************************/
+#[derive(Clone, Debug)]
+pub struct CreateUUIDGasParameters {
+    pub base: InternalGas,
+}
+
+fn native_create_uuid(
+    gas_params: &CreateUUIDGasParameters,
+    context: &mut SafeNativeContext,
+    mut _ty_args: Vec<Type>,
+    _args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    context.charge(gas_params.base)?;
+
+    let mut transaction_context = context
+        .extensions_mut()
+        .get_mut::<NativeTransactionContext>();
+    transaction_context.uuid_counter += 1;
+
+    let hash_vec = AuthenticationKeyPreimage::uuid(TransactionDerivedUUID {
+        txn_hash: transaction_context.txn_hash.clone(),
+        uuid_counter: transaction_context.uuid_counter,
+    });
```

Comment:
Here, you want to call `AuthenticationKey::from_preimage` over the returned `AuthenticationKeyPreimage`.

Reply:
Done.